### PR TITLE
 Rework Modules CMake to be (more) idiomatic.

### PR DESCRIPTION
**Pull request by EricWF**
- Fix bug in documentation regarding dependencies.
- Rework Modules CMake to be (more) idiomatic.

### CODE REVIEW

File: libcxx/modules/CMakeLists.txt.in [hunk 1]
```diff
@@ -50,10 +50,15 @@ endif()
...
     -nostdinc++
+    @LIBCXX_COMPILE_FLAGS@
+)
+target_compile_options(std
+  PRIVATE
     -Wno-reserved-module-identifier
     -Wno-reserved-user-defined-literal
-    @LIBCXX_COMPILE_FLAGS@
 )
+target_link_options(std PUBLIC -nostdlib++ -Wl,-rpath,${LIBCXX_BUILD}/lib/x86_64-unknown-linux-gnu/ -L${LIBCXX_BUILD}/lib/x86_64-unknown-linux-gnu)
```
Comment (mordante) [inline libcxx/modules/CMakeLists.txt.in:60]: This has the target `x86_64-unknown-linux-gnu` hard-coded. I expect this to fail on other platforms.

File: libcxx/modules/CMakeLists.txt.in [hunk 2]
```diff
@@ -50,10 +50,15 @@ endif()
 target_compile_options(std
   PUBLIC
     -nostdinc++
+    @LIBCXX_COMPILE_FLAGS@
+)
+target_compile_options(std
+  PRIVATE
     -Wno-reserved-module-identifier
     -Wno-reserved-user-defined-literal
-    @LIBCXX_COMPILE_FLAGS@
```
Comment (mordante) [inline libcxx/modules/CMakeLists.txt.in:55]: Why is this removed?

File: libcxx/docs/Modules.rst [hunk 3]
```diff
@@ -178,34 +178,13 @@ This is a small sample program that uses the module ``std``. It consists of a
...
-  # Adjust project compiler flags
-  #
-
-  add_compile_options($<$<COMPILE_LANGUAGE:CXX>:-fprebuilt-module-path=${std_BINARY_DIR}/CMakeFiles/std.dir/>)
-  add_compile_options($<$<COMPILE_LANGUAGE:CXX>:-fprebuilt-module-path=${std_BINARY_DIR}/CMakeFiles/std.compat.dir/>)
-  add_compile_options($<$<COMPILE_LANGUAGE:CXX>:-nostdinc++>)
-  # The include path needs to be set to be able to use macros from headers.
-  # For example from, the headers <cassert> and <version>.
-  add_compile_options($<$<COMPILE_LANGUAGE:CXX>:-isystem>)
-  add_compile_options($<$<COMPILE_LANGUAGE:CXX>:${LIBCXX_BUILD}/include/c++/v1>)
```
Comment (mordante) [inline libcxx/docs/Modules.rst:191]: Why is this no longer needed?

**Review by mordante — CHANGES_REQUESTED** [review]
Thanks for this improvement. Does the current version still work when using modules from the build directory?

File: libcxx/docs/Modules.rst (hunk 3 above)
Comment (EricWF) [reply libcxx/docs/Modules.rst:191]: The module cache path bits aren't needed because CMake adds `-fmodule-file=std=<path-to-build>.pcm`, so that's magically handled behind the scenes. 

The remainder aren't needed because when you link to a target, the flags of that target are added to your flag set.

This added those flags globally to every target, but the cmake way to do this is via the PUBLIC compile/link options.

**Comment by EricWF** [comment]
> Thanks for this improvement. Does the current version still work when using modules from the build directory?

It should. though now the generated cmake lists isn't dependent on LIBCXX_BUILD, and instead just hardcodes all of the paths to the build directory.

Ideally we would just have the user link and build against a system libc++, so all that's needed is `-stdlib=libc++`

File: libcxx/modules/CMakeLists.txt.in (hunk 1 above)
Comment (EricWF) [reply libcxx/modules/CMakeLists.txt.in]: Fixed.

**Review by mordante — APPROVED** [review]
Thanks this looks like a nice simplification! I've tested it locally and it works.
LGTM!

**Comment by EricWF** [comment]
I think the next step is to actually build a module during the CMake build
